File: tests/perf/02_csdid_perf.py
"""Track C performance -- Callaway-Sant'Anna staggered DiD.

Task (both sides): group-time ATTs by outcome regression against the
never-treated, analytic SEs, the joint pre-trend test, and the simple and
dynamic aggregations, on the shared input ``data/02_csdid_<units>.csv``
(periods 1..5, cohorts 2, 3, 4; ``_data.py``), one thread. StatsPAI's
``sp.callaway_santanna`` returns all four in one call; the reference runs
``did::att_gt`` (which computes the pre-test) and two ``did::aggte`` calls
(``02_csdid_perf.R``). The simple ATT and its SE are recorded outside the
timed call.
"""

from __future__ import annotations

from _common import TimingResult, time_repeat, write_results
from _data import CSDID_T, SIZES, load

import statspai as sp

EST = "02_csdid"
N_REPS = 3


def main() -> None:
    rows: list[TimingResult] = []
    for n_units in SIZES[EST]:
        df, digest = load(EST, n_units)

        def run_sp():
            return sp.callaway_santanna(
                df,
                y="y",
                g="first_treat",
                t="year",
                i="unit",
                estimator="reg",
                control_group="nevertreated",
            )

        fit = run_sp()
        assert fit.model_info.get("event_study") is not None
        med, iqr, mn, mx, peak = time_repeat(run_sp, n_reps=N_REPS, warmup=1)
        rows.append(
            TimingResult(
                estimator=EST,
                side="py",
                n=int(n_units * CSDID_T),
                n_reps=N_REPS,
                median_time_s=med,
                iqr_time_s=iqr,
                min_time_s=mn,
                max_time_s=mx,
                peak_mem_mb=peak,
                extra={
                    "n_units": n_units,
                    "T": CSDID_T,
                    "data_sha256": digest,
                    "task": "ATT(g,t) reg/never-treated, analytic SE, pre-test, "
                    "simple + dynamic aggregation",
                    "estimate": float(fit.estimate),
                    "se": float(fit.se),
                },
            )
        )
        print(f"  N_units={n_units:>6}  median={med:.3f}s")
    write_results(EST, "py", rows)


if __name__ == "__main__":
    main()

File: tests/perf/02_csdid_perf.R
# Track C performance -- CS-DiD (R side).
# Same task and bytes as 02_csdid_perf.py: att_gt (reg, never-treated,
# analytic SEs; computes the pre-test) + aggte simple + aggte dynamic on
# data/02_csdid_<units>.csv (periods 1..5, cohorts 2, 3, 4).
.args <- commandArgs(trailingOnly = FALSE)
.f <- grep("^--file=", .args, value = TRUE)
source(file.path(if (length(.f)) dirname(normalizePath(sub("^--file=", "", .f[1]))) else getwd(), "_common.R"))
suppressPackageStartupMessages(library(did))

EST <- "02_csdid"
N_REPS <- 3L
T_PERIODS <- 5L
rows <- list()
for (n_units in c(1000L, 5000L, 25000L)) {
  d <- read_perf_data(EST, n_units)
  df <- d$df
  # did recodes never-treated first_treat = 0 to Inf inside a data.table; on
  # an integer column that assignment becomes NA and those units are dropped
  # silently ("No never-treated group is available"), so the fit uses the
  # last cohort as controls. read.csv parses the column as integer. The
  # previous version of this benchmark built it with integer() and timed
  # that degenerate problem; the output check in compare_perf.py caught it.
  df$first_treat <- as.numeric(df$first_treat)
  fn <- function() {
    fit <- did::att_gt(
      yname = "y", tname = "year", idname = "unit", gname = "first_treat",
      data = df, control_group = "nevertreated", est_method = "reg",
      bstrap = FALSE, cband = FALSE
    )
    list(
      simple = did::aggte(fit, type = "simple", bstrap = FALSE, cband = FALSE),
      dynamic = did::aggte(fit, type = "dynamic", bstrap = FALSE, cband = FALSE),
      pretest = fit$Wpval
    )
  }
  out <- fn()
  res <- time_one(fn, n_reps = N_REPS)
  rows[[length(rows) + 1L]] <- perf_row(EST, n_units * T_PERIODS, N_REPS, res, list(
    n_units = n_units, T = T_PERIODS,
    data_sha256 = d$sha256,
    task = "ATT(g,t) reg/never-treated, analytic SE, pre-test, simple + dynamic aggregation",
    estimate = out$simple$overall.att,
    se = out$simple$overall.se,
    package = "did"
  ))
  cat(sprintf("  N_units=%6d  median=%.3fs\n", n_units, res$median))
}
write_perf(EST, rows, "did")
